### PR TITLE
Jed/ns dmplex

### PR DESCRIPTION
**Pull request by jedbrown**
Major refactor of Navier-Stokes solver, with DMPlex support superseding the legacy structured grid code. Includes a basic automated test and support for convergence studies (presently with advection). This includes lots of work from @LeilaGhaffari and @jrwrigh.

This branch has conflicts, but a tentative merge to master is available in https://github.com/CEED/libCEED/tree/jed/ns-dmplex-merge. Once approved, I'll amend the commit message and merge to master.

### CODE REVIEW

**Comment by jedbrown** [comment]
Note that if you use `-viz_refine` for refined visualization of high-order basis functions, you'll also need [this MR](https://gitlab.com/petsc/petsc/-/merge_requests/2515) in PETSc, which should land soon. I believe everything else will work with moderately recent PETSc.

File: examples/navier-stokes/.gitignore [hunk 1]
```diff
@@ -1,4 +1,6 @@
 navierstokes
+nsplex
```
Comment (jeremylt) [inline examples/navier-stokes/.gitignore]: Not needed since we are renaming.

File: examples/petsc/multigrid.c [hunk 2]
```diff
@@ -318,6 +318,7 @@ int main(int argc, char **argv) {
 
     // Multiplicity
     CeedElemRestrictionGetMultiplicity(ceeddata[i]->Erestrictu,
+                                       CEED_NOTRANSPOSE,
```
Comment (jeremylt) [inline examples/petsc/multigrid.c]: `CEED_NOTRANSPOSE` argument is no longer part of this function

File: examples/navier-stokes/.gitignore (hunk 1 above)
Comment (valeriabarra) [reply examples/navier-stokes/.gitignore]: Updated

File: examples/petsc/multigrid.c (hunk 2 above)
Comment (jeremylt) [reply examples/petsc/multigrid.c]: Ah, the proposed rebase got this

Comment (jedbrown) [reply examples/petsc/multigrid.c]: Thanks. I'd already fixed this in my merge branch (since it's a semantic conflict). Should we be reviewing there?

Comment (valeriabarra) [reply examples/petsc/multigrid.c]: Ok. You can cherry-pick the small commits I've done here then

Comment (jedbrown) [reply examples/petsc/multigrid.c]: Nah, your changes are fine. I'd planned to repeat the merge. (Git has a clever feature called [rerere](https://git-scm.com/docs/git-rerere) that makes such things manageable.)

**Comment by valeriabarra** [comment]
What are the configurations I should test? I have tested the `density_current` problem and I don't see what we used to see before

**Comment by valeriabarra** [comment]
I tried to run `density_current` with the `-implicit` option but it aborts with `EXIT CODE: 9`

**Comment by valeriabarra** [comment]
> What are the configurations I should test? I have tested the `density_current` problem and I don't see what we used to see before

See some explicit test outputs [here](https://drive.google.com/drive/folders/1Bn5KJTH0S1FahtcrObxG-FlfWfTdDERZ?usp=sharing) and [here](https://drive.google.com/open?id=1BCtNBP79VpinEN2x-C28ieGq6FdpZY5J)

**Comment by LeilaGhaffari** [comment]
> I tried to run `density_current` with the `-implicit` option but it aborts with `EXIT CODE: 9`

This should work. It's running on my laptop now but it is very slow. 

```
./navierstokes -ts_monitor -petscspace_degree 3 -dm_plex_box_faces 8,8,4 -ts_max_time 2 -dm_view -implicit -ksp_atol 1e-3 -snes_atol 1e-1 -ts_adapt_atol 2e-2 -ts_view -stab none -problem density_current -snes_monitor -ksp_converged_reason -snes_fd_color -ts_adapt_monitor
```

**Comment by LeilaGhaffari** [comment]
Also, `density_current_primitive` is not functional yet. @jedbrown , are you planning to merge it anyway?

**Comment by LeilaGhaffari** [comment]
> > What are the configurations I should test? I have tested the `density_current` problem and I don't see what we used to see before
> 
> See some explicit test outputs [here](https://drive.google.com/drive/folders/1Bn5KJTH0S1FahtcrObxG-FlfWfTdDERZ?usp=sharing) and [here](https://drive.google.com/open?id=1BCtNBP79VpinEN2x-C28ieGq6FdpZY5J)

Thank you for sharing them. Could you also share the configurations for these tests?

**Comment by valeriabarra** [comment]
> > I tried to run `density_current` with the `-implicit` option but it aborts with `EXIT CODE: 9`
> 
> This should work. It's running on my laptop now but it is very slow.
> 
> ```
> ./navierstokes -ts_monitor -petscspace_degree 3 -dm_plex_box_faces 8,8,4 -ts_max_time 2 -dm_view -implicit -ksp_atol 1e-3 -snes_atol 1e-1 -ts_adapt_atol 2e-2 -ts_view -stab none -problem density_current -snes_monitor -ksp_converged_reason -snes_fd_color -ts_adapt_monitor
> ```

Thanks. I had just tried `-implicit` without any other time stepper or snes options. If some options are "must-haves", it should be documented.

**Comment by valeriabarra** [comment]
> > > What are the configurations I should test? I have tested the `density_current` problem and I don't see what we used to see before
> > 
> > 
> > See some explicit test outputs [here](https://drive.google.com/drive/folders/1Bn5KJTH0S1FahtcrObxG-FlfWfTdDERZ?usp=sharing) and [here](https://drive.google.com/open?id=1BCtNBP79VpinEN2x-C28ieGq6FdpZY5J)
> 
> Thank you for sharing them. Could you also share the configurations for these tests?

@LeilaGhaffari I have just dropped `param*.txt` files with the parameters used for each simulation. They were meant to reproduce some of the old results we have, so I have used the same parameters (except the number of boxes for resolution)

File: examples/navier-stokes/navierstokes.c [hunk 3]
```diff
@@ -291,24 +347,46 @@ static PetscErrorCode TSMonitor_NS(TS ts, PetscInt stepno, PetscReal time,
...
-  ierr = VecScatterBegin(user->ltog, Q, user->Qloc, INSERT_VALUES,
-                         SCATTER_REVERSE); CHKERRQ(ierr);
-  ierr = VecScatterEnd(user->ltog, Q, user->Qloc, INSERT_VALUES,
-                       SCATTER_REVERSE); CHKERRQ(ierr);
-  ierr = MapLocalToDMGlobal(user, user->Qloc, U, scale); CHKERRQ(ierr);
+  ierr = DMGetLocalVector(user->dm, &Qloc);CHKERRQ(ierr);
+  ierr = PetscObjectSetName((PetscObject)Qloc, "StateVec"); CHKERRQ(ierr);
+  ierr = VecZeroEntries(Qloc);CHKERRQ(ierr);
+  ierr = DMGlobalToLocal(user->dm, Q, INSERT_VALUES, Qloc); CHKERRQ(ierr);
+  if (0) {
```
Comment (valeriabarra) [inline examples/navier-stokes/navierstokes.c]: Do we want to keep this?

**Comment by jedbrown** [comment]
Latest version seems to produce stable and reasonable (if poorly resolved still) vortex behavior in density current. Note that this is with explicit integration and no stabilization.
```
./navierstokes -petscspace_degree 2 -problem density_current -dm_plex_box_faces 64,1,32 -ly 125 -bc_wall 1,2,5,6 -bc_slip_y 3,4 -ts_adapt_monitor -ts_adapt_safety 0.7
```

image

File: examples/navier-stokes/navierstokes.c (hunk 3 above)
Comment (jedbrown) [reply examples/navier-stokes/navierstokes.c]: It's a longer term issue. I made this placeholder so I'll delete that dead code from the branch. https://gitlab.com/petsc/petsc/issues/564

**Comment by jedbrown** [comment]
The necessary PETSc fixes have merged to master, so things like slip can be tested now. The following is an implicit run with SUPG.
image
Performance of SUPG should be more readily testable in bubble and vortex transport experiments that we can run after merging. @valeriabarra Do you have further concerns?

This run (not tuned):
```
./navierstokes -petscspace_degree 2 -qextra 2 -problem density_current -dm_plex_box_faces 20,1,20 -lx 4000 -ly 200 -lz
 4000 -bc_wall 1,2,5,6 -bc_slip_y 3,4 -ts_adapt_monitor -ts_adapt_safety 0.5 -implicit -snes_mf -stab supg -snes_atol 1e-3 -ksp_atol 1e-4 -ksp_rtol 1e-2
```

**Comment by valeriabarra** [comment]
> The necessary PETSc fixes have merged to master, so things like slip can be tested now. The following is an implicit run with SUPG.
> image
> Performance of SUPG should be more readily testable in bubble and vortex transport experiments that we can run after merging. @valeriabarra Do you have further concerns?
> 
> This run (not tuned):
> 
> ```
> ./navierstokes -petscspace_degree 2 -qextra 2 -problem density_current -dm_plex_box_faces 20,1,20 -lx 4000 -ly 200 -lz
>  4000 -bc_wall 1,2,5,6 -bc_slip_y 3,4 -ts_adapt_monitor -ts_adapt_safety 0.5 -implicit -snes_mf -stab supg -snes_atol 1e-3 -ksp_atol 1e-4 -ksp_rtol 1e-2
> ```

Thanks for the update. My only concern is that we are retro-compatible with our old results and that any new capability/feature introduced is well documented in the READMEs and in the code, to allow for easy testing. I'm going to test this further and let you know if I find anything odd.

File: examples/navier-stokes/common.h [hunk 4]
```diff
@@ -113,6 +169,39 @@ CEED_QFUNCTION(Setup)(void *ctx, CeedInt Q,
   return 0;
 }
 
+CEED_QFUNCTION(Setup2d)(void *ctx, CeedInt Q,
+                      const CeedScalar *const *in, CeedScalar *const *out) {
+  // Inputs
+  const CeedScalar (*J)[2][Q] = (CeedScalar(*)[2][Q])in[0],
```
Comment (valeriabarra) [inline examples/navier-stokes/common.h]: Do we care about making this declaration and the other similar ones as 
```
const CeedScalar (*J)[2][Q] = (const CeedScalar(*)[2][Q])in[0]
```
to avoid the warnings on intel compilers?

Comment (jedbrown) [reply examples/navier-stokes/common.h]: Better to be strict to avoid the warnings (even if they aren't useful). Are there many such instances? (GCC used to warn, but decided to switch to a more useful interpretation of the standard.)

Comment (valeriabarra) [reply examples/navier-stokes/common.h]: Oh yes, there are many. We have a lot of these declarations using VLA in our header files:
```
In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(46):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/common.h(120): warning #144: a value of type "CeedScalar={double} (*)[3][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[3][Q]"
    const CeedScalar (*J)[3][Q] = (CeedScalar(*)[3][Q])in[0],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(46):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/common.h(175): warning #144: a value of type "CeedScalar={double} (*)[2][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[2][Q]"
    const CeedScalar (*J)[2][Q] = (CeedScalar(*)[2][Q])in[0],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(46):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/common.h(220): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*u)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(107): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*X)[Q] = (CeedScalar(*)[Q])in[0];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(150): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(151): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(152): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[2],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(153): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*x)[Q] = (CeedScalar(*)[Q])in[3];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(267): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(268): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(269): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdot)[Q] = (CeedScalar(*)[Q])in[2],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(47):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection.h(270): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[3];
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(89): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*X)[Q] = (CeedScalar(*)[Q])in[0];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(132): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(133): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(134): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[2],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(135): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*x)[Q] = (CeedScalar(*)[Q])in[3];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(224): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(225): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(226): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdot)[Q] = (CeedScalar(*)[Q])in[2],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(48):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/advection2d.h(227): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[3];
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(132): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*X)[Q] = (CeedScalar(*)[Q])in[0];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(209): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(210): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(211): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[2],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(212): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*x)[Q] = (CeedScalar(*)[Q])in[3];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(461): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(462): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(463): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdot)[Q] = (CeedScalar(*)[Q])in[2],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(49):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent.h(464): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[3];
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(134): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*X)[Q] = (CeedScalar(*)[Q])in[0];
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(213): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
    const CeedScalar (*q)[Q] = (CeedScalar(*)[Q])in[0],
                               ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(214): warning #144: a value of type "CeedScalar={double} (*)[5][Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[5][Q]"
                     (*dq)[5][Q] = (CeedScalar(*)[5][Q])in[1],
                                   ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(215): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdot)[Q] = (CeedScalar(*)[Q])in[2],
                                  ^

In file included from /projects/vaba3353/CEED/libCEED/examples/navier-stokes/navierstokes.c(50):
/projects/vaba3353/CEED/libCEED/examples/navier-stokes/densitycurrent_primitive.h(216): warning #144: a value of type "CeedScalar={double} (*)[Q]" cannot be used to initialize an entity of type "const CeedScalar={double} (*)[Q]"
                     (*qdata)[Q] = (CeedScalar(*)[Q])in[3];
                                   ^


```

**Comment by jedbrown** [comment]
I updated the merge branch (https://github.com/CEED/libCEED/tree/jed/ns-dmplex-merge) and think this is ready. I think it's premature to carefully document interface options (beyond `-help` and judicious comments) before we have #446 merged). We need to have a fair amount of churn in this example to accommodate new boundary conditions and related features, but I'd rather that be a follow-up. We do need small, fast-running tests so that each future refactoring isn't a labor-intensive tour of facilities and eyeball checks.

Note that there is an unrelated Travis-CI issue due to this breaking commit: https://github.com/libocca/occa/commit/e3dfd526b5556231e2ba3e225f848802114dd4da

**Comment by valeriabarra** [comment]
I'm updating the README with the help strings. 
I'm also double checking sone unused variable warnings that I see in my IDE

**Comment by jedbrown** [comment]
Thanks for your updates, @valeriabarra.

With OCCA fixed upstream, the remaining test failure is due to our new wall boundary condition (zero mass and energy flux), which were not supported in the release version of PETSc that we're using (a bug in DMPlex). Shall I do full-Dirichlet conditions (at least for the test that runs automatically) or update to newer PETSc to support the physical boundary conditions? The next release of PETSc should come in about a month and there are more things I want to get into it (that we'll use in the coming months).

File: .gitignore [hunk 5]
```diff
@@ -22,3 +23,10 @@ config.mk
 
 # Debugging
 vgcore*
+*.json
+
+# Editor Misc
```
Comment (valeriabarra) [inline .gitignore]: Are these Editor misc for a particular editor? Should we keep them?

Comment (jedbrown) [reply .gitignore]: They're popular editors and it's pretty harmless to list here, versus having them accidentally committed or everyone having to add them to `.git/info/exclude`.

Comment (jrwrigh) [reply .gitignore]: the `.vscode` folder is for VSCode. It holds miscellaneous project settings for VSCode.
`.history` is a specific [VSCode extension](https://marketplace.visualstudio.com/items?itemName=xyz.local-history) that stores copies of files whenever they're saved.

`.ccls-cache` is for the C language server, [CCLS](https://github.com/MaskRay/ccls). CCLS can be used for any editor that supports the Language Server Protocol (which in my case is both VSCode and vim). 
`.vim` is for vim, but I've only ever seen it exist when using CCLS inside vim.

As for whether to keep them, I'm not sure. I just put them there on my forked branch because I was tired of seeing 20+ files I didn't care about whenever I ran `git status`.

Comment (valeriabarra) [reply .gitignore]: Ok thanks. The reason why I was asking is that I'm revising this to fix the merge conflicts in `user-manual`

Comment (jedbrown) [reply .gitignore]: Cool; I'd prefer to keep these ignores. Basically zero cost and makes developer life easier.

Comment (valeriabarra) [reply .gitignore]: I have always refrained to adding the QtCreator ones. I thought I was the only one using it and did not pertain other users. Can I add those too?

Comment (jedbrown) [reply .gitignore]: Sure, go for it.

Comment (valeriabarra) [reply .gitignore]: Cool, thanks. I've updated it in the `user-manual` branch, since that file was touched for the merge conflicts anyways and the branch should be merged soon.
